pkg/river/token/builder/builder.go

```diff
@@ -0,0 +1,186 @@
+	for i, node := range b.nodes {
+		rawToks = append(rawToks, node.Tokens()...)
+
+		if i+1 < len(b.nodes) {
+			// Append a terminator between each statement in the Body.
+			rawToks = append(rawToks, Token{
+				Type: token.ILLEGAL,
+				Lit:  "\n",
+			})
+		}
+	}
+	return rawToks
+}
+
+// AppendTokens appens raw tokens to the Body.
+func (b *Body) AppendTokens(tokens []Token) {
+	b.nodes = append(b.nodes, tokensSlice(tokens))
+}
+
+// AppendBlock adds a new block inside of the Body.
+func (b *Body) AppendBlock(block *Block) {
+	b.nodes = append(b.nodes, block)
+}
+
+// SetAttributeTokens sets an attribute to the Body whose value is a set of raw
+// tokens. If the attribute was previously set, its value tokens are updated.
+//
+// Attributes will be written out in the order they were initially created.
+func (b *Body) SetAttributeTokens(name string, tokens []Token) {
+	attr := b.getOrCreateAttribute(name)
+	attr.RawTokens = tokens
+}
+
+func (b *Body) getOrCreateAttribute(name string) *attribute {
+	for _, n := range b.nodes {
+		if attr, ok := n.(*attribute); ok && attr.Name == name {
+			return attr
+		}
+	}
+
+	newAttr := &attribute{Name: name}
+	b.nodes = append(b.nodes, newAttr)
+	return newAttr
+}
+
+// SetAttributeValue sets an attribute in the Body whose value is converted
+// from a Go value to a River value. The Go value is encoded using the normal
+// Go to River encoding rules.
+//
+// If the attribute was previously set, its value tokens are updated.
+//
+// Attributes will be written out in the order they were initially crated.
+func (b *Body) SetAttributeValue(name string, goValue interface{}) {
+	attr := b.getOrCreateAttribute(name)
+	attr.RawTokens = tokenEncode(goValue)
+}
+
+type attribute struct {
+	Name      string
+	RawTokens []Token
+}
+
+func (attr *attribute) Tokens() []Token {
+	var toks []Token
+
+	toks = append(toks, Token{Type: token.IDENT, Lit: attr.Name})
+	toks = append(toks, Token{Type: token.ASSIGN})
+	toks = append(toks, attr.RawTokens...)
+
+	return toks
+}
+
+// A Block encapsulates a body within a named and labeled River block. Blocks
+// must be created by calling NewBlock, but its public struct fields may be
+// safely modified by callers.
+type Block struct {
+	// Public fields, safe to be changed by callers:
+
+	Name  []string
+	Label string
+
+	// Private fields:
+
+	body *Body
+}
+
+// NewBlock returns a new Block with the given name and label. The name/label
+// can be updated later by modifying the Block's public fields.
+func NewBlock(name []string, label string) *Block {
+	return &Block{
+		Name:  name,
+		Label: label,
+
+		body: newBody(),
+	}
+}
+
+// Tokens returns the File as a set of Tokens.
+func (b *Block) Tokens() []Token {
+	var toks []Token
+
+	for i, frag := range b.Name {
+		toks = append(toks, Token{Type: token.IDENT, Lit: frag})
+		if i+1 < len(b.Name) {
+			toks = append(toks, Token{Type: token.DOT})
+		}
+	}
+
+	toks = append(toks, Token{Type: token.ILLEGAL, Lit: " "})
```

Comment:
Same here and below, not sure I get the ILLEGAL part. Is it because it's just for presentation so it doesn't matter?

Reply:
See the Token docstring:

```
// A Token is a wrapper around token.Token which contains the token type
// alongside its literal. Use token.ILLEGAL to write literal characters such as
// whitespace.
```

We just unfortunately don't have a good way to say "this should be written as-is" other than re-using token.ILLEGAL.